docs/adding-an-endpoint.adoc

```diff
@@ -0,0 +1,59 @@
+* Either way, you should add reference the new resource when https://github.com/oxidecomputer/omicron/blob/1dfe47c1b3122bc4f32a9c517cb31b1600581ea2/nexus/src/authz/oso_generic.rs#L119-L148[constructing the Oso structure]
+
+=== **Application**
+* Add any "business logic" for the resource to xref:../nexus/src/app[the app directory]
+* This layer bridges the gap between the database and external services.
+* If your application logic involes any multi-step operations which would be interrupted by Nexus stopping mid-execution (due to reboot, crash, failure, etc), it is recommended to use a https://github.com/oxidecomputer/omicron/tree/1dfe47c1b3122bc4f32a9c517cb31b1600581ea2/nexus/src/app/sagas[saga] to define the operations durably.
+
+=== **Database**
+* `CREATE TABLE` for the resource in xref:../common/src/sql/dbinit.sql[dbinit.sql] (https://github.com/oxidecomputer/omicron/blob/1dfe47c1b3122bc4f32a9c517cb31b1600581ea2/common/src/sql/dbinit.sql#L1103-L1129[Example])
+* Add an equivalent schema for the resource in xref:../nexus/db-model/src/schema.rs[schema.rs], which allows https://docs.diesel.rs/master/diesel/index.html[Diesel] to translate raw SQL to rust queries (https://github.com/oxidecomputer/omicron/blob/1dfe47c1b3122bc4f32a9c517cb31b1600581ea2/nexus/db-model/src/schema.rs#L144-L155[Example])
+* Add a Rust representation of the database object to xref:../nexus/db-model/src[the DB model] (https://github.com/oxidecomputer/omicron/blob/1dfe47c1b3122bc4f32a9c517cb31b1600581ea2/nexus/db-model/src/ip_pool.rs#L24-L40[Example])
+** Remember to make the object visible; this is usually defined in xref:../nexus/db-model/src/lib.rs[lib.rs] (https://github.com/oxidecomputer/omicron/blob/1dfe47c1b3122bc4f32a9c517cb31b1600581ea2/nexus/db-model/src/lib.rs#L102[Example]).
+* Methods to send queries to the database are defined as part of the https://github.com/oxidecomputer/omicron/tree/1dfe47c1b3122bc4f32a9c517cb31b1600581ea2/nexus/src/db/datastore[datastore] structure (https://github.com/oxidecomputer/omicron/blob/1dfe47c1b3122bc4f32a9c517cb31b1600581ea2/nexus/src/db/datastore/ip_pool.rs[Example])
+
+=== **Testing**
+
+* Authorization
+** There exists a https://github.com/oxidecomputer/omicron/blob/main/nexus/src/authz/policy_test[policy test] which compares all Oso objects against an expected policy. New resources are usually added to https://github.com/oxidecomputer/omicron/blob/main/nexus/src/authz/policy_test/resources.rs[resources.rs] to get coverage.
+* openapi
+** Nexus generates a new openapi spec from the dropshot endpoints. If you modify endpoints, you'll usually need to `EXPECTORATE=overwrite cargo test` the following targets...
+*** External API: `test_nexus_openapi`
+*** Internal API: `test_nexus_openapi_internal`
```

Comment:
These tests are good, in that they're getting so much coverage, but they also kinda suck from a developer UX point of view. Testing any API changes feels like it requires re-running `cargo test` with and without `EXPECTORATE` a *bunch* of times, especially since changing the openapi file causes different code generation, which means the next call to `cargo test` might not actually build...

Reply:
Was it faster to just run it manually without going via the tests, e.g.

```console
$ cargo run -p omicron-nexus --bin nexus -- -I nexus/examples/config.toml > openapi/nexus-internal.json
$ cargo run -p omicron-nexus --bin nexus -- -O nexus/examples/config.toml > openapi/nexus.json
$ cargo run -p omicron-sled-agent --bin sled-agent -- openapi > openapi/sled-agent.json
```

Reply:
Definitely seems to build fewer packages, at least. I will be trying this. Normally I run

```
EXPECTORATE=overwrite cargo t integration_tests::commands::test_nexus_openapi
```

I only ever work on the external API.

Reply:
Added both the test and non-test options